### PR TITLE
add code samples to Hero page

### PR DESCRIPTION
**Pull request by sw-yx**
## What was changed

add code samples to Hero page. see sample here https://temporaldotio-pxbjugb5g-temporal.vercel.app/

CleanShot 2021-07-05 at 04 00 52@2x

### CODE REVIEW

**Comment by sw-yx** [comment]
cc @jh3y

File: codesamples/old-samples/sample.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+public class SampleWorkflow {
```
Comment (mfateev) [inline codesamples/old-samples/sample.java:1]: This is not valid Java SDK code.
WorkflowClient.start is used by outside code to start workflows. It is prohibited inside a workflow code.

Workflow.sleep is prohibited outside of the workflow code which is the case for the static main method.

File: codesamples/index.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+const path = require('path');
+
+const arr = ['js', 'go', 'java', 'php'];
+
+for (const lang of arr) {
+  shiki
+    .getHighlighter({
+      theme: 'material-palenight'
+    })
+    .then((highlighter) => {
```
Comment (bergundy) [inline codesamples/index.js:12]: Is this promise supposed to be awaited somewhere for error propagation?

File: codesamples/sample.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
+func mainWorkflow(ctx workflow.Context, userId string) error {
+	// ...
+	intervals := [3]int{1, 7, 30}
+	for _, i := range intervals {
+		time.Sleep(i * time.Day)                                           // sleep for days!
+		_ = workflow.ExecuteActivity(ctx, SendEmailActivity).Get(ctx, nil) // timeouts and retries!
```
Comment (bergundy) [inline codesamples/sample.go:6]: We should pass in the `userId` and `i` (interval) to the activity.
(Supposedly the interval can be used in order to customize the retention email).
Should we rename `SendEmailActivity` to `SendUserRetentionEmail` to clarify what this sample is about?

File: codesamples/sample.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+func mainWorkflow(ctx workflow.Context, userId string) error {
+	// ...
+	intervals := [3]int{1, 7, 30}
+	for _, i := range intervals {
```
Comment (bergundy) [inline codesamples/sample.go:4]: Let's rename `i` to `interval`

File: codesamples/sample.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+		time.Sleep(i * time.Day)                                           // sleep for days!
+		_ = workflow.ExecuteActivity(ctx, SendEmailActivity).Get(ctx, nil) // timeouts and retries!
+	}
+	// we can also signal or cancel the workflow as needed
```
Comment (bergundy) [inline codesamples/sample.go:8]: I haven't thought this through but we might want to say what "as needed" means.
In this sample we're hinting that we'd cancel the workflow for example when the user finally interacts with the system or changes their notification preferences.

File: codesamples/sample.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+int intervals[] = { 1, 7, 30 };
+for (int i : intervals) {   
+  Workflow.sleep(Duration.ofDays(i)); // sleep for days!
+  workflow.sendEmail(userId) // automatic timeouts + retries!
```
Comment (bergundy) [inline codesamples/sample.java:4]:
```suggestion
  activities.sendUserRetentionEmail(interval, userId) // automatic timeouts + retries!
```

File: codesamples/sample.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+async function main(userId) {
+  const intervals = [1, 7, 30]; // Nag user until giving up
+  for (const interval of intervals) {
+    await sleep(interval * DAYS);
+    await sendEmailActivity(interval, userId); // automatic timeouts + retries
```
Comment (bergundy) [inline codesamples/sample.js:5]: How about this?
```suggestion
    await activities.sendUserRetentionEmail(interval, userId); // automatic timeouts + retries
```

This flavor can be used with
```ts
import * as activities from '@activities';
```

File: codesamples/sample.php [hunk 8]
```diff
@@ -0,0 +1,8 @@
+public function main(string $userID) {
```
Comment (bergundy) [inline codesamples/sample.php:1]: Let's get someone who knows PHP to review this.

File: codesamples/sample.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+int intervals[] = { 1, 7, 30 };
```
Comment (bergundy) [inline codesamples/sample.java:1]: We should probably put this in a class method, the other samples are wrapped in a method.

File: codesamples/sample.go (hunk 5 above)
Comment (bergundy) [reply codesamples/sample.go:8]: This comment applies to the other samples too

File: codesamples/index.js (hunk 2 above)
Comment (sw-yx) [reply codesamples/index.js:12]: no it just runs... dont worry about this code

File: codesamples/sample.go (hunk 3 above)
Comment (sw-yx) [reply codesamples/sample.go:6]: like this?

```suggestion
		_ = workflow.ExecuteActivity(ctx, SendUserRetentionEmail, userId, i).Get(ctx, nil) // timeouts and retries!
```

File: codesamples/sample.go (hunk 4 above)
Comment (sw-yx) [reply codesamples/sample.go:4]:
```suggestion
	for _, interval := range intervals {
```

File: codesamples/sample.go (hunk 5 above)
Comment (sw-yx) [reply codesamples/sample.go:8]:
```suggestion
	// we can signal or cancel the workflow if the user does an action
```

File: codesamples/sample.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+
+public static class MainWorkflowImpl implements MainWorkflow {
+    public void PollUser(String[] args) {
```
Comment (tsurdilo) [inline codesamples/sample.java:3]: method name might be confusing heere

File: codesamples/sample.php (hunk 8 above)
Comment (tsurdilo) [reply codesamples/sample.php:1]: For Java workflow we should the class definition for this you could enclose the code in:

            class MainWorkflow implements MainWorkflowInterface
            {
            ...
            }

also function does not have to be "main" i think in all language examples the method name should make some sense (as in what it does) if possible

File: codesamples/sample.php [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+    $intervals = array(1, 7, 30);
+    foreach ($intervals as &$interval) {
+        yield Workflow::timer($interval * DAYS); // sleep for days!
+        yield $this->user->sendEmail($interval, $userID); // timeouts + retries!
```
Comment (tsurdilo) [inline codesamples/sample.php:5]: don't think there is need for the "yield" here.

Comment (tsurdilo) [reply codesamples/sample.php:5]: also $this->user seems as this is a workflow field
in other examples like java its made more clear that this is actually an activity exec

maybe change the code to

         $this->userActivity->sendEmail($interval, $userID);

**Comment by tsurdilo** [comment]
I don't see much value in the samples.xyz code, sorry. 
If you would like to keep them at least make the time intervals dynamic, for example workflow input rather than being hard coded. Also the comments below the method  mention that "you can signal...", which imo is confusing as we use workflow sleep and should in that case use for example workflow away, with the time duration and an "exit" condition which can be triggered by signal. Just my 2c

File: codesamples/sample.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+
+public static class MainWorkflowImpl implements MainWorkflow {
```
Comment (tsurdilo) [inline codesamples/sample.java:2]: Maybe we should have a workflow class with a better name, like maybe "UserEngagementWorkflow" or idk what the overarching workflow from this examples is really doing to come up with a better name.

File: codesamples/sample.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+    public void PollUser(String[] args) {
+      int intervals[] = { 1, 7, 30 };
+      for (int interval: intervals) {   
+        Workflow.sleep(Duration.ofDays(interval)); // sleep for days!
+        activities.sendUserRetentionEmail(interval, userId) // automatic timeouts + retries!
```
Comment (tsurdilo) [inline codesamples/sample.java:7]: I think that (for all samples) the "automatic timeouts + retried!" comment is imo unclear if its for the activity invocation, or the sending of email/retention emails. Maybe change to 
// Activities are retried by default. You can set Activity execution timeouts!
something like that so its more clear?

File: codesamples/sample.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+async function main(userId) {
+  const intervals = [1, 7, 30]; // Nag user until giving up
+  for (const interval of intervals) {
+    await sleep(interval * DAYS);
```
Comment (tsurdilo) [inline codesamples/sample.js:4]: i think this sample is missing the "sleep for days!" comment as in other examples

File: codesamples/sample.js [hunk 15]
```diff
@@ -0,0 +1,8 @@
+async function main(userId) {
+  const intervals = [1, 7, 30]; // Nag user until giving up
```
Comment (tsurdilo) [inline codesamples/sample.js:2]: "Nag user until giving up" - is that the equivalent of 
"Send user retention email after 1, 8 and 38 days"?

File: codesamples/sample.java (hunk 13 above)
Comment (sw-yx) [reply codesamples/sample.java:7]: will do

File: codesamples/sample.java (hunk 12 above)
Comment (sw-yx) [reply codesamples/sample.java:2]: will do

File: codesamples/sample.php (hunk 11 above)
Comment (sw-yx) [reply codesamples/sample.php:5]: got it, thank you

File: codesamples/sample.java (hunk 10 above)
Comment (sw-yx) [reply codesamples/sample.java:3]: will change

File: codesamples/sample.js (hunk 15 above)
Comment (sw-yx) [reply codesamples/sample.js:2]: good point

**Comment by sw-yx** [comment]
i think we would all like to add back the signals in a sensible way at some point - this is where i stopped with the signals samples: https://gist.github.com/sw-yx/4b8d660635ede59d8870b17b6b4e9e95

i would personally love to showcase them and `Workflow.await` which i think is amazing - but we will ship this for now and iterate. thank you for the reviews everyone.

**Comment by sw-yx** [comment]
please use #37 to continue providing feedback on the samples

**Pull request merged**